L1Trigger/L1THGCal/plugins/backend/HGCalHistoClusteringWrapper.cc

```diff
@@ -0,0 +1,201 @@
+#include "L1Trigger/L1THGCal/interface/backend/HGCalSeed_SA.h"
+#include "L1Trigger/L1THGCal/interface/backend/HGCalMulticluster_SA.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
+#include "L1Trigger/L1THGCal/interface/backend/HGCalShowerShape.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerTools.h"
+#include "L1Trigger/L1THGCal/interface/backend/HGCalTriggerClusterIdentificationBase.h"
+
+class HGCalHistoClusteringWrapper : public HGCalHistoClusteringWrapperBase {
+public:
+  HGCalHistoClusteringWrapper(const edm::ParameterSet& conf);
+  ~HGCalHistoClusteringWrapper() override {}
+
+  void configure(const std::pair<const edm::EventSetup&, const edm::ParameterSet&>& configuration) override;
+
+  void process(const std::pair<const std::vector<edm::Ptr<l1t::HGCalCluster>>,
+                               const std::vector<std::pair<GlobalPoint, double>>>& inputClustersAndSeeds,
+               std::pair<l1t::HGCalMulticlusterBxCollection&, l1t::HGCalClusterBxCollection&>&
+                   outputMulticlustersAndRejectedClusters) const override;
+
+private:
+  void convertCMSSWInputs(const std::vector<edm::Ptr<l1t::HGCalCluster>>& clustersPtrs,
+                          l1thgcfirmware::HGCalClusterSACollection& clusters_SA,
+                          const std::vector<std::pair<GlobalPoint, double>>& seeds,
+                          l1thgcfirmware::HGCalSeedSACollection& seeds_SA) const;
+  void convertAlgorithmOutputs(const l1thgcfirmware::HGCalMulticlusterSACollection& multiclusters_out,
+                               const l1thgcfirmware::HGCalClusterSACollection& rejected_clusters_out,
+                               const std::vector<edm::Ptr<l1t::HGCalCluster>>& clustersPtrs,
+                               l1t::HGCalMulticlusterBxCollection& multiclusters,
+                               l1t::HGCalClusterBxCollection& rejected_clusters) const;
+
+  void clusterizeHisto(const l1thgcfirmware::HGCalClusterSACollection& inputClusters,
+                       const l1thgcfirmware::HGCalSeedSACollection& inputSeeds,
+                       l1thgcfirmware::HGCalMulticlusterSACollection& outputMulticlusters,
+                       l1thgcfirmware::HGCalClusterSACollection& outputRejectedClusters) const;
+
+  void eventSetup(const edm::EventSetup& es) {
+    triggerTools_.eventSetup(es);
+    shape_.eventSetup(es);
+    if ((!dr_byLayer_coefficientA_.empty() && (dr_byLayer_coefficientA_.size() - 1) < triggerTools_.lastLayerBH()) ||
+        (!dr_byLayer_coefficientB_.empty() && (dr_byLayer_coefficientB_.size() - 1) < triggerTools_.lastLayerBH())) {
+      throw cms::Exception("Configuration")
+          << "The per-layer dR values go up to " << (dr_byLayer_coefficientA_.size() - 1) << "(A) and "
+          << (dr_byLayer_coefficientB_.size() - 1) << "(B), while layers go up to " << triggerTools_.lastLayerBH()
+          << "\n";
+    }
```

Comment:
Rather than checking these `dr_byLayer_coeff*` it might be better to check those inside the config, since it is those ones which are used in the end.
And I'm actually wondering if it is useful to store these coefficients in the wrapper, as they are not used here. Having them only in the config should be enough.

Reply:
Right, actually I was thinking we can keep the checks in the wrapper, but access the coeffs through the available public getters of the config.